File: src/opentimelineio/composition.cpp
// SPDX-License-Identifier: Apache-2.0
// Copyright Contributors to the OpenTimelineIO project

#include "opentimelineio/composition.h"
#include "opentimelineio/clip.h"
#include "opentimelineio/vectorIndexing.h"

#include <assert.h>
#include <set>

namespace opentimelineio { namespace OPENTIMELINEIO_VERSION {

Composition::Composition(
    std::string const&              name,
    std::optional<TimeRange> const& source_range,
    AnyDictionary const&            metadata,
    std::vector<Effect*> const&     effects,
    std::vector<Marker*> const&     markers)
    : Parent(name, source_range, metadata, effects, markers)
{}

Composition::~Composition()
{
    clear_children();
}

std::string
Composition::composition_kind() const
{
    static std::string kind = "Composition";
    return kind;
}

void
Composition::clear_children()
{
    for (Composable* child: _children)
    {
        child->_set_parent(nullptr);
    }

    _children.clear();
    _child_set.clear();
}

bool
Composition::set_children(
    std::vector<Composable*> const& children,
    ErrorStatus*                    error_status)
{
    for (auto child: children)
    {
        if (child->parent())
        {
            if (error_status)
            {
                *error_status = ErrorStatus::CHILD_ALREADY_PARENTED;
            }
            return false;
        }
    }

    for (auto child: children)
    {
        child->_set_parent(this);
    }

    _children  = decltype(_children)(children.begin(), children.end());
    _child_set = std::set<Composable*>(children.begin(), children.end());
    return true;
}

bool
Composition::insert_child(
    int          index,
    Composable*  child,
    ErrorStatus* error_status)
{
    if (child->parent())
    {
        if (error_status)
        {
            *error_status = ErrorStatus::CHILD_ALREADY_PARENTED;
        }
        return false;
    }

    child->_set_parent(this);

    index = adjusted_vector_index(index, _children);
    if (index >= int(_children.size()))
    {
        _children.emplace_back(child);
    }
    else
    {
        _children.insert(_children.begin() + std::max(index, 0), child);
    }

    _child_set.insert(child);
    return true;
}

bool
Composition::set_child(int index, Composable* child, ErrorStatus* error_status)
{
    index = adjusted_vector_index(index, _children);
    if (index < 0 || index >= int(_children.size()))
    {
        if (error_status)
        {
            *error_status = ErrorStatus::ILLEGAL_INDEX;
        }
        return false;
    }

    if (_children[index] != child)
    {
        if (child->parent())
        {
            if (error_status)
            {
                *error_status = ErrorStatus::CHILD_ALREADY_PARENTED;
            }
            return false;
        }

        _children[index]->_set_parent(nullptr);
        _child_set.erase(_children[index]);
        child->_set_parent(this);
        _children[index] = child;
        _child_set.insert(child);
    }
    return true;
}

bool
Composition::remove_child(int index, ErrorStatus* error_status)
{
    if (_children.empty())
    {
        if (error_status)
        {
            *error_status = ErrorStatus::ILLEGAL_INDEX;
        }
        return false;
    }

    index = adjusted_vector_index(index, _children);

    _child_set.erase(_children[index]);

    if (size_t(index) >= _children.size())
    {
        _children.back()->_set_parent(nullptr);
        _children.pop_back();
    }
    else
    {
        index = std::max(index, 0);
        _children[index]->_set_parent(nullptr);
        _children.erase(_children.begin() + index);
    }

    return true;
}

int
Composition::index_of_child(Composable const* child, ErrorStatus* error_status)
    const
{
    for (size_t i = 0; i < _children.size(); i++)
    {
        if (_children[i] == child)
        {
            return int(i);
        }
    }

    if (error_status)
    {
        *error_status                = ErrorStatus::NOT_A_CHILD_OF;
        error_status->object_details = this;
    }
    return -1;
}

bool
Composition::read_from(Reader& reader)
{
    if (reader.read("children", &_children) && Parent::read_from(reader))
    {
        for (Composable* child: _children)
        {
            if (!child->_set_parent(this))
            {
                reader.error(ErrorStatus::CHILD_ALREADY_PARENTED);
                return false;
            }
        }
    }
    return true;
}

void
Composition::write_to(Writer& writer) const
{
    Parent::write_to(writer);
    writer.write("children", _children);
}

bool
Composition::is_parent_of(Composable const* other) const
{
    Composition const* cur_parent = other->_parent;
    if (cur_parent == this)
        return true;

    std::set<Composition const*> visited;
    while (cur_parent && visited.count(cur_parent) == 0)
    {
        if (cur_parent == this)
            return true;

        visited.insert(cur_parent);
        cur_parent = cur_parent->_parent;
    }
    return false;
}

std::pair<std::optional<RationalTime>, std::optional<RationalTime>>
Composition::handles_of_child(
    Composable const* /* child */,
    ErrorStatus* /* error_status */) const
{
    return std::make_pair(
        std::optional<RationalTime>(),
        std::optional<RationalTime>());
}

std::vector<Composition*>
Composition::_path_from_child(
    Composable const* child,
    ErrorStatus*      error_status) const
{
    auto                      current = child->parent();
    std::vector<Composition*> parents{ current };

    while (current != this)
    {
        current = current->parent();
        if (!current)
        {
            if (error_status)
            {
                *error_status                = ErrorStatus::NOT_DESCENDED_FROM;
                error_status->object_details = this;
            }
            return parents;
        }
        parents.push_back(current);
    }

    return parents;
}

TimeRange
Composition::range_of_child_at_index(int /* index */, ErrorStatus* error_status)
    const
{
    if (error_status)
    {
        *error_status = ErrorStatus::NOT_IMPLEMENTED;
    }
    return TimeRange();
}

TimeRange
Composition::trimmed_range_of_child_at_index(
    int /* index */,
    ErrorStatus* error_status) const
{
    if (error_status)
    {
        *error_status = ErrorStatus::NOT_IMPLEMENTED;
    }
    return TimeRange();
}

std::map<Composable*, TimeRange>
Composition::range_of_all_children(ErrorStatus* error_status) const
{
    if (error_status)
    {
        *error_status = ErrorStatus::NOT_IMPLEMENTED;
    }
    return std::map<Composable*, TimeRange>();
}

// XXX should have reference_space argument or something
TimeRange
Composition::range_of_child(Composable const* child, ErrorStatus* error_status)
    const
{
    auto parents = _path_from_child(child, error_status);
    if (is_error(error_status))
    {
        return TimeRange();
    }

    Composition const*       reference_space = this; // XXX
    std::optional<TimeRange> result_range;
    auto                     current = child;

    assert(!parents.empty());
    for (auto parent: parents)
    {
        const int index = parent->index_of_child(current, error_status);
        if (is_error(error_status))
        {
            return TimeRange();
        }

        auto parent_range =
            parent->range_of_child_at_index(index, error_status);
        if (is_error(error_status))
        {
            return TimeRange();
        }

        if (!result_range)
        {
            result_range = parent_range;
            current      = parent;
            continue;
        }

        result_range = TimeRange(
            result_range->start_time() + parent_range.start_time(),
            std::min(result_range->duration(), parent_range.duration()));
        current = parent;
    }

    return (reference_space != this) ? transformed_time_range(
                                           *result_range,
                                           reference_space,
                                           error_status)
                                     : *result_range;
}

// XXX should have reference_space argument or something
std::optional<TimeRange>
Composition::trimmed_range_of_child(
    Composable const* child,
    ErrorStatus*      error_status) const
{
    auto parents = _path_from_child(child, error_status);
    if (is_error(error_status))
    {
        return TimeRange();
    }

    std::optional<TimeRange> result_range;
    auto                     current = child;

    assert(!parents.empty());
    for (auto parent: parents)
    {
        const int index = parent->index_of_child(current, error_status);
        if (is_error(error_status))
        {
            return TimeRange();
        }

        auto parent_range =
            parent->trimmed_range_of_child_at_index(index, error_status);
        if (is_error(error_status))
        {
            return TimeRange();
        }

        if (!result_range)
        {
            result_range = parent_range;
            current      = parent;
            continue;
        }

        result_range = TimeRange(
            result_range->start_time() + parent_range.start_time(),
            std::min(result_range->duration(), parent_range.duration()));
    }

    if (!source_range())
    {
        return result_range;
    }

    auto new_start_time =
        std::max(source_range()->start_time(), result_range->start_time());
    if (new_start_time > result_range->end_time_exclusive())
    {
        return std::nullopt;
    }

    auto new_duration = std::min(
                            result_range->end_time_exclusive(),
                            source_range()->end_time_exclusive())
                        - new_start_time;
    if (new_duration.value() < 0)
    {
        return std::nullopt;
    }

    return TimeRange(new_start_time, new_duration);
}

std::vector<Composable*>
Composition::_children_at_time(RationalTime t, ErrorStatus* error_status) const
{
    std::vector<Composable*> result;

    // range_of_child_at_index is O(i), so this loop is quadratic:
    for (size_t i = 0; i < _children.size() && !is_error(error_status); i++)
    {
        if (range_of_child_at_index(int(i), error_status).contains(t))
        {
            result.push_back(_children[i]);
        }
    }

    return result;
}

std::optional<TimeRange>
Composition::trim_child_range(TimeRange child_range) const
{
    if (!source_range())
    {
        return child_range;
    }

    const TimeRange sr = *source_range();
    bool past_end_time = sr.start_time() >= child_range.end_time_exclusive();
    bool before_start_time =
        sr.end_time_exclusive() <= child_range.start_time();

    if (past_end_time || before_start_time)
    {
        return std::nullopt;
    }

    if (child_range.start_time() < sr.start_time())
    {
        child_range = TimeRange::range_from_start_end_time(
            sr.start_time(),
            child_range.end_time_exclusive());
    }

    if (child_range.end_time_exclusive() > sr.end_time_exclusive())
    {
        child_range = TimeRange::range_from_start_end_time(
            child_range.start_time(),
            sr.end_time_exclusive());
    }

    return child_range;
}

bool
Composition::has_child(Composable* child) const
{
    return _child_set.find(child) != _child_set.end();
}

SerializableObject::Retainer<Composable>
Composition::child_at_time(
    RationalTime const& search_time,
    ErrorStatus*        error_status,
    bool                shallow_search) const
{
    Retainer<Composable> result;

    auto range_map = range_of_all_children(error_status);
    if (is_error(error_status))
    {
        return result;
    }

    // find the first item whose end_time_exclusive is after the
    const auto first_inside_range = _bisect_left(
        search_time,
        [&range_map](Composable* child) {
            return range_map[child].end_time_exclusive();
        },
        error_status);
    if (is_error(error_status))
    {
        return result;
    }

    // find the last item whose start_time is before the
    const auto last_in_range = _bisect_right(
        search_time,
        [&range_map](Composable* child) {
            return range_map[child].start_time();
        },
        error_status,
        first_inside_range);
    if (is_error(error_status))
    {
        return result;
    }

    // limit the search to children who are in the search_range
    std::vector<Retainer<Composable>> possible_matches;
    for (auto child = _children.begin() + first_inside_range;
         child < _children.begin() + last_in_range;
         ++child)
    {
        possible_matches.push_back(child->value);
    }
    for (const auto& thing: possible_matches)
    {
        if (range_map[thing].overlaps(search_time))
        {
            result = thing;
            break;
        }
    }

    // if the search cannot or should not continue
    auto composition =
        Retainer<Composition>(dynamic_cast<Composition*>(result.value));
    if (!result || shallow_search || !composition)
    {
        return result;
    }

    // before you recurse, you have to transform the time into the
    // space of the child
    const auto child_search_time =
        transformed_time(search_time, composition.value, error_status);
    if (is_error(error_status))
    {
        return result;
    }

    result = composition.value->child_at_time(
        child_search_time,
        error_status,
        shallow_search);
    if (is_error(error_status))
    {
        return result;
    }
    return result;
}

std::vector<SerializableObject::Retainer<Composable>>
Composition::children_in_range(
    TimeRange const& search_range,
    ErrorStatus*     error_status) const
{
    std::vector<Retainer<Composable>> children;

    auto range_map = range_of_all_children(error_status);
    if (is_error(error_status))
    {
        return children;
    }

    // find the first item whose end_time_inclusive is after the
    // start_time of the search range
    const auto first_inside_range = _bisect_left(
        search_range.start_time(),
        [&range_map](Composable* child) {
            return range_map[child].end_time_inclusive();
        },
        error_status);
    if (is_error(error_status))
    {
        return children;
    }

    // find the last item whose start_time is before the
    // end_time_inclusive of the search_range
    const auto last_in_range = _bisect_right(
        search_range.end_time_inclusive(),
        [&range_map](Composable* child) {
            return range_map[child].start_time();
        },
        error_status,
        first_inside_range);
    if (is_error(error_status))
    {
        return children;
    }

    // limit the search to children who are in the search_range
    for (auto child = _children.begin() + first_inside_range;
         child < _children.begin() + last_in_range;
         ++child)
    {
        children.push_back(child->value);
    }
    return children;
}

int64_t
Composition::_bisect_right(
    RationalTime const&                             tgt,
    std::function<RationalTime(Composable*)> const& key_func,
    ErrorStatus*                                    error_status,
    std::optional<int64_t>                          lower_search_bound,
    std::optional<int64_t>                          upper_search_bound) const
{
    if (*lower_search_bound < 0)
    {
        if (error_status)
        {
            *error_status = ErrorStatus(
                ErrorStatus::Outcome::INTERNAL_ERROR,
                "lower_search_bound must be non-negative");
        }
        return 0;
    }
    if (!upper_search_bound)
    {
        upper_search_bound = std::optional<int64_t>(_children.size());
    }
    int64_t midpoint_index = 0;
    while (*lower_search_bound < *upper_search_bound)
    {
        midpoint_index = static_cast<int64_t>(
            std::floor((*lower_search_bound + *upper_search_bound) / 2.0));

        if (tgt < key_func(_children[midpoint_index]))
        {
            upper_search_bound = midpoint_index;
        }
        else
        {
            lower_search_bound = midpoint_index + 1;
        }
    }

    return *lower_search_bound;
}

int64_t
Composition::_bisect_left(
    RationalTime const&                             tgt,
    std::function<RationalTime(Composable*)> const& key_func,
    ErrorStatus*                                    error_status,
    std::optional<int64_t>                          lower_search_bound,
    std::optional<int64_t>                          upper_search_bound) const
{
    if (*lower_search_bound < 0)
    {
        if (error_status)
        {
            *error_status = ErrorStatus(
                ErrorStatus::Outcome::INTERNAL_ERROR,
                "lower_search_bound must be non-negative");
        }
        return 0;
    }
    if (!upper_search_bound)
    {
        upper_search_bound = std::optional<int64_t>(_children.size());
    }
    int64_t midpoint_index = 0;
    while (*lower_search_bound < *upper_search_bound)
    {
        midpoint_index = static_cast<int64_t>(
            std::floor((*lower_search_bound + *upper_search_bound) / 2.0));

        if (key_func(_children[midpoint_index]) < tgt)
        {
            lower_search_bound = midpoint_index + 1;
        }
        else
        {
            upper_search_bound = midpoint_index;
        }
    }

    return *lower_search_bound;
}

bool
Composition::has_clips() const
{
    for (auto child: children())
    {
        if (dynamic_cast<Clip*>(child.value))
        {
            return true;
        }
        else if (auto child_comp = dynamic_cast<Composition*>(child.value))
        {
            if (child_comp->has_clips())
            {
                return true;
            }
        }
    }
    return false;
}

}} // namespace opentimelineio::OPENTIMELINEIO_VERSION

File: src/java-opentimelineio/src/main/java/io/opentimeline/opentimelineio/Composition.java
package io.opentimeline.opentimelineio;

import io.opentimeline.OTIONative;
import io.opentimeline.opentime.RationalTime;
import io.opentimeline.opentime.TimeRange;
import io.opentimeline.util.Pair;

import java.util.*;
import java.util.function.Function;
import java.util.stream.Collectors;
import java.util.stream.Stream;

/**
 * Base class for an OTIO Item that contains other Items.
 * Should be subclassed (for example by Track and Stack), not used directly.
 */
public class Composition extends Item {

    protected Composition() {
    }

    Composition(OTIONative otioNative) {
        this.nativeManager = otioNative;
    }

    public Composition(
            String name,
            TimeRange sourceRange,
            AnyDictionary metadata,
            List<Effect> effects,
            List<Marker> markers) {
        this.initObject(
                name,
                sourceRange,
                metadata,
                effects,
                markers);
    }

    public Composition(Composition.CompositionBuilder builder) {
        this.initObject(
                builder.name,
                builder.sourceRange,
                builder.metadata,
                builder.effects,
                builder.markers);
    }

    private void initObject(String name,
                            TimeRange sourceRange,
                            AnyDictionary metadata,
                            List<Effect> effects,
                            List<Marker> markers) {
        Effect[] effectsArray = new Effect[effects.size()];
        effectsArray = effects.toArray(effectsArray);
        Marker[] markersArray = new Marker[markers.size()];
        markersArray = markers.toArray(markersArray);
        this.initialize(
                name,
                sourceRange,
                metadata,
                effectsArray,
                markersArray);
        this.nativeManager.className = this.getClass().getCanonicalName();
    }

    private native void initialize(String name,
                                   TimeRange sourceRange,
                                   AnyDictionary metadata,
                                   Effect[] effects,
                                   Marker[] markers);

    public static class CompositionBuilder {
        private String name = "";
        private TimeRange sourceRange = null;
        private AnyDictionary metadata = new AnyDictionary();
        private List<Effect> effects = new ArrayList<>();
        private List<Marker> markers = new ArrayList<>();

        public CompositionBuilder() {
        }

        public Composition.CompositionBuilder setName(String name) {
            this.name = name;
            return this;
        }

        public Composition.CompositionBuilder setSourceRange(TimeRange sourceRange) {
            this.sourceRange = sourceRange;
            return this;
        }

        public Composition.CompositionBuilder setMetadata(AnyDictionary metadata) {
            this.metadata = metadata;
            return this;
        }

        public Composition.CompositionBuilder setEffects(List<Effect> effects) {
            this.effects = effects;
            return this;
        }

        public Composition.CompositionBuilder setMarkers(List<Marker> markers) {
            this.markers = markers;
            return this;
        }

        public Composition build() {
            return new Composition(this);
        }
    }

    /**
     * Returns a label specifying the kind of composition.
     *
     * @return the composition kind
     */
    public native String getCompositionKind();

    /**
     * @return all the chidlren held by this Composition.
     */
    public List<Composable> getChildren() {
        return Arrays.asList(getChildrenNative());
    }

    private native Composable[] getChildrenNative();

    /**
     * Remove all children from the composition and clear their parents.
     */
    public native void clearChildren();

    /**
     * Set children for this Composition.
     *
     * @param children    a list of Composables
     * @param errorStatus errorStatus to report an error if any child cannot be added.
     */
    public void setChildren(List<Composable> children, ErrorStatus errorStatus) {
        Composable[] childrenArray = new Composable[children.size()];
        childrenArray = children.toArray(childrenArray);
        setChildrenNative(childrenArray, errorStatus);
    }

    private native void setChildrenNative(Composable[] children, ErrorStatus errorStatus);

    /**
     * Insert a child Composable at an index.
     *
     * @param index       index at which child is to be inserted.
     * @param child       child Composable to be inserted.
     * @param errorStatus errorStatus to report an error if child cannot be inserted.
     * @return was the child inserted successfully?
     */
    public native boolean insertChild(int index, Composable child, ErrorStatus errorStatus);

    /**
     * Set the child at a particular index. The needs to exist for the child to be set.
     *
     * @param index       index to set the child
     * @param child       child Composable to be set
     * @param errorStatus errorStatus to report an error if child cannot be set.
     * @return was the child set successfully?
     */
    public native boolean setChild(int index, Composable child, ErrorStatus errorStatus);

    /**
     * Remove the child at any index.
     *
     * @param index       index from which the child needs to be removed.
     * @param errorStatus errorStatus to report an error if child cannot be removed.
     * @return was the child removed successfully?
     */
    public native boolean removeChild(int index, ErrorStatus errorStatus);

    /**
     * Append the child to the end of the Composition.
     *
     * @param child       child to be appended
     * @param errorStatus errorStatus to report an error if child cannot be appended.
     * @return was the child appended successfully?
     */
    public native boolean appendChild(Composable child, ErrorStatus errorStatus);

    /**
     * @param composable Composable to check ancestry of
     * @return true if this is a parent or ancestor of the composable.
     */
    public native boolean isParentOf(Composable composable);

    /**
     * If media beyond the ends of this child are visible due to adjacent
     * Transitions (only applicable in a Track) then this will return the
     * head and tail offsets as a Pair of RationalTime objects. If no handles
     * are present on either side, then null is returned instead of a
     * RationalTime.
     *
     * @param child       child Composable to get handles
     * @param errorStatus errorStatus to report error while fetching handles
     * @return head and tail offsets as a Pair of RationalTime objects
     */
    public native Pair<RationalTime, RationalTime> getHandlesOfChild(
            Composable child, ErrorStatus errorStatus);

    /**
     * Return the range of a child item in the time range of this composition.
     * For example, with a track:
     * [ClipA][ClipB][ClipC]
     * this.getRangeOfChildAtIndex(2, errorStatus) will return:
     * TimeRange(ClipA.duration + ClipB.duration, ClipC.duration)
     *
     * @param index       index of child whose range is to be fetched
     * @param errorStatus errorStatus to report error if range cannot be fetched
     * @return range of child at index
     */
    public native TimeRange getRangeOfChildAtIndex(int index, ErrorStatus errorStatus);

    /**
     * Return the trimmed range of the child item at index in the time
     * range of this composition.
     * <p>
     * For example, with a track:
     * [     ]
     * [ClipA][ClipB][ClipC]
     * The range of index 2 (ClipC) will be just like
     * getRangeOfChildAtIndex() but trimmed based on this Composition's
     * sourceRange.
     *
     * @param index       index of child whose trimmed range is to be fetched
     * @param errorStatus errorStatus to report error if trimmed range cannot be fetched
     * @return trimmed range of child at index
     */
    public native TimeRange getTrimmedRangeOfChildAtIndex(int index, ErrorStatus errorStatus);

    /**
     * The range of the child not trimmed based on this composition's sourceRange.
     * For example:
     * |     [-----]     | seq
     * [-----------------] Clip A
     * If ClipA has duration 17, and seq has sourceRange: 5, duration 15,
     * seq.getRangeOfChild(Clip A) will return (0, 17)
     * ignoring the sourceRange of seq.
     * <p>
     * To get the range of the child with the sourceRange applied, use the
     * getTrimmedRangeOfChild() method.
     *
     * @param child       child Composable whose range is to be fetched
     * @param errorStatus errorStatus to report error if range cannot be fetched
     * @return range of the child not trimmed based on this composition's sourceRange
     */
    public native TimeRange getRangeOfChild(Composable child, ErrorStatus errorStatus);

    /**
     * Get range of the child, after the sourceRange is applied.
     * Example
     * |     [-----]     | seq
     * [-----------------] Clip A
     * If ClipA has duration 17, and seq has sourceRange: 5, duration 10,
     * seq.getTrimmedRangeOfChild(Clip A) will return (5, 10)
     * Which is trimming the range according to the sourceRange of seq.
     * To get the range of the child without the sourceRange applied, use the
     * getRangeOfChild() method.
     * Another example
     * |  [-----]   | seq sourceRange starts on frame 4 and goes to frame 8
     * [ClipA][ClipB] (each 6 frames long)
     * <p>
     * seq.getRangeOfChild(ClipA) = 0, duration 6
     * seq.getTrimmedRangeOfChild(ClipB) = 4, duration 2
     *
     * @param child       child Composable whose trimmed range is to be fetched
     * @param errorStatus errorStatus to report error if range cannot be fetched
     * @return range of the child trimmed based on this composition's sourceRange
     */
    public native TimeRange getTrimmedRangeOfChild(Composable child, ErrorStatus errorStatus);

    /**
     * Trim childRange to this.sourceRange
     *
     * @param childRange range to be trimmed
     * @return null if childRange has no intersection with this.sourceRange, else trimmed range
     */
    public native TimeRange trimChildRange(TimeRange childRange);

    /**
     * @param child child to find
     * @return does the composition contain the child?
     */
    public native boolean hasChild(Composable child);

    /**
     * Return a HashMap mapping children to their range in this object.
     *
     * @param errorStatus errorStatus to report any error while fetching ranges
     * @return a HashMap mapping children to their range in this object
     */
    public native HashMap<Composable, TimeRange> getRangeOfAllChildren(ErrorStatus errorStatus);

    /**
     * Return the child that overlaps with time searchTime.
     * searchTime is in the space of self.
     * If shallowSearch is false, will recurse into compositions.
     *
     * @param searchTime    the time at which the child is to be fetched
     * @param shallowSearch should the algorithm recurse into compositions or not?
     * @param errorStatus   errorStatus to report any error while fetching the child
     * @return the child that overlaps with time searchTime
     */
    public Composable getChildAtTime(RationalTime searchTime, boolean shallowSearch, ErrorStatus errorStatus) {
        HashMap<Composable, TimeRange> rangeOfAllChildren = this.getRangeOfAllChildren(errorStatus);
        List<Composable> children = this.getChildren();
        // find the first item whose endTimeExclusive is after the target
        int firstInsideRange = bisectLeft(
                children,
                searchTime,
                child -> rangeOfAllChildren.get(child).endTimeExclusive(),
                0, null);
        // find the last item whose startTime is before the target
        int lastInRange = bisectRight(
                children,
                searchTime,
                child -> rangeOfAllChildren.get(child).getStartTime(),
                firstInsideRange, null);
        // limit the search to children who are in the searchRange
        children = children.subList(firstInsideRange, lastInRange);
        Composable result = null;
        for (Composable potentialMatch : children) {
            if (rangeOfAllChildren.get(potentialMatch).overlaps(searchTime)) {
                result = potentialMatch;
                break;
            }
        }
        // if the search cannot or should not continue
        if (shallowSearch || !(result instanceof Composition))
            return result;

        // before you recurse, you have to transform the time into the
        // space of the child
        RationalTime childSearchTime = this.getTransformedTime(searchTime, (Composition) result, errorStatus);
        if (errorStatus.getOutcome() != ErrorStatus.Outcome.OK) throw new RuntimeException();
        return ((Composition) result).getChildAtTime(childSearchTime, shallowSearch, errorStatus);
    }

    /**
     * Return the child that overlaps with time searchTime.
     * searchTime is in the space of self.
     * This function will recurse into compositions.
     *
     * @param searchTime  the time at which the child is to be fetched
     * @param errorStatus errorStatus to report any error while fetching the child
     * @return the child that overlaps with time searchTime
     */
    public Composable getChildAtTime(RationalTime searchTime, ErrorStatus errorStatus) {
        return this.getChildAtTime(searchTime, false, errorStatus);
    }

    /**
     * Stream that returns each child of specified type contained in the composition in
     * the order in which it is found.
     *
     * @param searchRange   if not null, only children whose range overlaps with the search range will be in the stream.
     * @param descendedFrom only children who are a descendent of the descendedFrom type will be in the stream
     * @param shallowSearch should the algorithm recurse into compositions or not?
     * @param errorStatus   errorStatus to report any error while iterating
     * @param <T>           type of children to fetch
     * @return a Stream consisting of all the children of specified type in the composition in the order in which it is found
     */
    public <T extends Composable> Stream<T> eachChild(
            TimeRange searchRange, Class<T> descendedFrom, boolean shallowSearch, ErrorStatus errorStatus) {
        List<Composable> children;
        if (searchRange != null) {
            HashMap<Composable, TimeRange> rangeOfAllChildren = this.getRangeOfAllChildren(errorStatus);
            children = this.getChildren();
            // find the first item whose endTimeInclusive is after the
            // startTime of the searchRange
            int firstInsideRange = bisectLeft(
                    children,
                    searchRange.getStartTime(),
                    child -> rangeOfAllChildren.get(child).endTimeInclusive(),
                    0, null);
            // find the last item whose startTime is before the
            // endTimeInclusive of the searchRange
            int lastInRange = bisectRight(
                    children,
                    searchRange.endTimeInclusive(),
                    child -> rangeOfAllChildren.get(child).getStartTime(),
                    firstInsideRange, null);
            children = children.subList(firstInsideRange, lastInRange);
        } else {
            children = this.getChildren();
        }

        return children.stream()
                .flatMap(element -> {
                            Stream<T> currentElementStream = Stream.empty();
                            if (descendedFrom.isAssignableFrom(element.getClass()))
                                currentElementStream = Stream.concat(Stream.of(descendedFrom.cast(element)), currentElementStream);
                            Stream<T> nestedStream = Stream.empty();
                            if (!shallowSearch && element instanceof Composition) {
                                nestedStream = ((Composition) element).eachChild(
                                        searchRange == null ? null : this.getTransformedTimeRange(searchRange, ((Composition) element), errorStatus),
                                        descendedFrom,
                                        shallowSearch,
                                        errorStatus);
                            }
                            return Stream.concat(currentElementStream, nestedStream);
                        }
                );
    }

    /**
     * Stream that returns each child contained in the composition in
     * the order in which it is found.
     *
     * @param searchRange   if not null, only children whose range overlaps with the search range will be in the stream.
     * @param shallowSearch should the algorithm recurse into compositions or not?
     * @param errorStatus   errorStatus to report any error while iterating
     * @return a Stream consisting of all the children in the composition (in the searchRange) in the order in which it is found
     */
    public Stream<Composable> eachChild(TimeRange searchRange, boolean shallowSearch, ErrorStatus errorStatus) {
        return eachChild(searchRange, Composable.class, shallowSearch, errorStatus);
    }

    /**
     * Stream that returns each child of specified type contained in the composition in
     * the order in which it is found. This stream will recurse into compositions.
     *
     * @param descendedFrom only children who are a descendent of the descendedFrom type will be in the stream
     * @param errorStatus   errorStatus to report any error while iterating
     * @param <T>           type of children to fetch
     * @return a Stream consisting of all the children of specified type in the composition in the order in which it is found
     */
    public <T extends Composable> Stream<T> eachChild(Class<T> descendedFrom, ErrorStatus errorStatus) {
        return eachChild(null, descendedFrom, false, errorStatus);
    }

    /**
     * Return the index of the last item in seqChildren such that all e in seqChildren[:index]
     * have keyFunction(e) <= target, and all e in seq[index:] have keyFunction(e) > target.
     * <p>
     * Thus, seqChildren.insert(index, value) will insert value after the rightmost item
     * such that meets the above condition.
     * <p>
     * lowerSearchBound and upperSearchBound bound the range of elements to be searched.
     * <p>
     * Assumes that seqChildren is already sorted.
     */
    private int bisectRight(
            List<Composable> seqChildren,
            RationalTime target,
            Function<Composable, RationalTime> keyFunction,
            Integer lowerSearchBound,
            Integer upperSearchBound) {
        if (lowerSearchBound == null)
            lowerSearchBound = 0;
        if (lowerSearchBound < 0)
            throw new IndexOutOfBoundsException("lowerSearchBound must be non-negative");
        if (upperSearchBound == null)
            upperSearchBound = seqChildren.size();

        while (lowerSearchBound < upperSearchBound) {
            int midPointIndex = Math.floorDiv(lowerSearchBound + upperSearchBound, 2);

            if (target.compareTo(keyFunction.apply(seqChildren.get(midPointIndex))) < 0) {
                upperSearchBound = midPointIndex;
            } else {
                lowerSearchBound = midPointIndex + 1;
            }
        }
        return lowerSearchBound;
    }

    /**
     * Return the index of the last item in seqChildren such that all e in seqChildren[:index]
     * have keyFunction(e) < target, and all e in seqChildren[index:] have keyFunction(e) >= target.
     * <p>
     * Thus, seqChildren.insert(index, value) will insert value before the leftmost item
     * such that meets the above condition.
     * <p>
     * lowerSearchBound and upperSearchBound bound the range of elements to be searched.
     * <p>
     * Assumes that seqChildren is already sorted.
     */
    private int bisectLeft(
            List<Composable> seqChildren,
            RationalTime target,
            Function<Composable, RationalTime> keyFunction,
            Integer lowerSearchBound,
            Integer upperSearchBound) {
        if (lowerSearchBound == null)
            lowerSearchBound = 0;
        if (lowerSearchBound < 0)
            throw new IndexOutOfBoundsException("lowerSearchBound must be non-negative");
        if (upperSearchBound == null)
            upperSearchBound = seqChildren.size();

        while (lowerSearchBound < upperSearchBound) {
            int midPointIndex = Math.floorDiv(lowerSearchBound + upperSearchBound, 2);

            if (keyFunction.apply(seqChildren.get(midPointIndex)).compareTo(target) < 0) {
                lowerSearchBound = midPointIndex + 1;
            } else {
                upperSearchBound = midPointIndex;
            }
        }
        return lowerSearchBound;
    }

    @Override
    public String toString() {
        return this.getClass().getCanonicalName() +
                "(" +
                "name=" + this.getName() +
                ", children=[" + this.getChildren()
                .stream().map(Objects::toString).collect(Collectors.joining(", ")) + "]" +
                ", sourceRange=" + this.getSourceRange() +
                ", metadata=" + this.getMetadata() +
                ")";
    }
}
